File: BOJ/2110_BinarySearch.cpp
#include <iostream>
#include <vector>
#include <algorithm>

// BOJ 2110 공유기 설치, Binary Search, 실버 1 
using namespace std;

int findGap(vector<int> house, int N, int C){
    int answer;
    int minGap = 1, maxGap = house[house.size()-1] - house[0];
    int mid;

    while (minGap <= maxGap){
        mid = (minGap + maxGap)/2;
        int router = 1;
        int last = 0, next = 1;

        while (next<N && router<C){
            if (house[next]>= house[last]+mid){
                router++;
                last = next;
                next = last+1;
            }else{
                next++;
            }
        }

        if (router < C){
            maxGap = mid-1;
        }else if (router == C){
            answer = mid;
            minGap = mid+1;
        }else{
            minGap = mid+1;
        }
    }

    return answer;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(NULL);

    int N, C;
    cin>>N>>C;

    vector<int> house(N, 0);
    for (int i = 0; i < N; ++i) {
        cin>>house[i];
    }

    sort(house.begin(), house.end());

    int answer = findGap(house, N, C);

    cout<<answer;

    return 0;
}

File: BOJ/2110_BinarySearch.swift
import Foundation
// boj 2110 공유기 설치, 이분 탐색, 골드 5 

let str1 = readLine()!.split(separator: " ").map{Int($0)!}
let N = str1[0]
let C = str1[1]


var temp = [Int]()
for _ in 0..<N {
    let time = Int(readLine()!)!
    temp.append(time)
}

var house = temp.sorted(by: < )

var left = 1
var right = house[N-1] - house[0]

var answer = 0

while left <= right {
    let mid = (left + right) / 2
    
    var router = 1
    var from = house[0]
    
    for i in 1..<N {
        if house[i]-from >= mid {
            router += 1
            from = house[i]
        }
    }
    
    if router>=C {
        left = mid + 1
        answer = max(answer, mid)
    } else {
        right = mid - 1
    }
}

print(answer)
